relay/circuit-v2.md

```diff
@@ -365,15 +365,15 @@ The wire representation is canonicalized, where elements of the message are writ
 ## Protobuf
 
 ```protobuf
-syntax = "proto2";
+syntax = "proto3";
 message HopMessage {
   enum Type {
     RESERVE = 0;
     CONNECT = 1;
     STATUS = 2;
   }
 
-  required Type type = 1;
+  Type type = 1;
```

Comment:
Are we expecting this to be backwards compatible with existing implementations that are expecting `proto2`?

If so this will break them when the enum value is `0` since it will be omitted on the wire as it's now a `singular` field and the rules are different.

Singular: https://developers.google.com/protocol-buffers/docs/proto3#specifying_field_rules

> It will be serialized to the wire unless it is the default value. 

Interestingly Google's protobuf.js gets this wrong - https://github.com/protobufjs/protobuf.js/issues/1822

Parsers following the older `proto2` schema will error as by their definition the field is `required` which means it needs to be present on the wire.

The only way for that to happen with a `proto3` definition is to mark the field `optional` and, at the application level, ensure it has a value set.

Reply:
@ckousik can you add some fuzzing tests to https://github.com/ckousik/proto2to3? The tests should have caught that.

Reply:
@MarcoPolo If we consider a case where we try to send `HopMessage { type: RESERVE }`, it should serialize to an empty buffer. I wrote a test for proto2 to decode the empty buffer and it passes. https://github.com/ckousik/proto2to3/blob/master/src/main.rs#L117

Reply:
Ah. Prost does not error here, although it should. See https://github.com/tokio-rs/prost#field-modifiers

Go would error here.

Reply:
Two options:
1. Change all formerly required fields to `optional` fields so that serializers **always** serialize the value. Try to find any other cases of incompatibility. And prove that we are backwards compatible.
2. Keep proto2 and have protons do the correct thing.

2 is the preferred option I think it lets us avoid the compatibility problem completely. But I recognize that it's not easy to do. And little value if protocols from now on are going to use proto3. How did this work before? Most everything so far has been proto2.

1 Makes sense if we can prove that we won't have any compatibility problems. As @achingbrain pointed out, this is tricky since implementations can do things that aren't technically what the protobuf spec requires (error on missing required field for example).

A suggestion to prove 1 would be to have both go and rust generate random valid messages (fuzzing) with proto3 and make sure the proto2 implementation can correctly parse those messages. If we have this I'm okay with option 1.

I know this is a bit annoying and hard. But so is backwards compatibility :)

Reply:
I'll do 1. and post an update on #509

Reply:
I've updated this PR.

Reply:
> I've run the experiments in [MarcoPolo/proto3-and-2-compat-experiments](https://github.com/MarcoPolo/proto3-and-2-compat-experiments). Could you update the PR to use this proto3 file? [MarcoPolo/proto3-and-2-compat-experiments@`main`/rust/src/circuitv2proto3.proto](https://github.com/MarcoPolo/proto3-and-2-compat-experiments/blob/main/rust/src/circuitv2proto3.proto?rgh-link-date=2023-02-01T02%3A15%3A34Z)

We are in the middle of transitioning to `quick-protobuf`. Let me add that to your experiments before you merge here please.